### PR TITLE
Enable `-Wshorten-64-to-32` warning if available.

### PR DESCRIPTION
**Pull request by edgchen1**
### Description

- Fix some warnings from Xcode build (`-Wshorten-64-to-32`).
- Enable `-Wshorten-64-to-32` warning if available.
- Some clean up in build.py including setting CMake generator more consistently.

Note: The branch name edgchen1/use_xcode_in_mac_ci_build is misleading - it reflects the initial approach that I tried.

### Motivation and Context

Fix warnings found in local Xcode build and ensure that they get caught by CI builds in the future.

### CODE REVIEW

File: cmake/onnxruntime_config.h.in [hunk 1]
```diff
@@ -3,23 +3,24 @@
...
 #cmakedefine HAS_FORMAT_TRUNCATION
-#cmakedefine HAS_BITWISE_INSTEAD_OF_LOGICAL
+#cmakedefine HAS_IGNORED_ATTRIBUTES
+#cmakedefine HAS_MAYBE_UNINITIALIZED
+#cmakedefine HAS_MISSING_BRACES
+#cmakedefine HAS_NONNULL_COMPARE
+#cmakedefine HAS_PARENTHESES
 #cmakedefine HAS_REALLOCARRAY
-#cmakedefine ORT_VERSION u8"@ORT_VERSION@"
+#cmakedefine HAS_SHORTEN_64_TO_32
```
Comment (edgchen1) [inline cmake/onnxruntime_config.h.in:19]: sorted these lines and added this one

File: tools/ci_build/build.py [hunk 2]
```diff
@@ -1183,19 +1184,6 @@ def generate_build_tree(
 
     if is_macOS() and not args.android:
         cmake_args += ["-DCMAKE_OSX_ARCHITECTURES=" + args.osx_arch]
-        if args.use_xcode:
```
Comment (edgchen1) [inline tools/ci_build/build.py:1186]: this check seems irrelevant now since our minimum cmake version higher than the ones handled here.

File: onnxruntime/core/framework/stream_execution_context.h [hunk 3]
```diff
@@ -37,7 +37,7 @@ class StreamExecutionContext {
    public:
     CountDownBarrier() : v_{0} {};
 
-    void Set(int32_t v) {
+    void Set(int_fast32_t v) {
```
Comment (skottmckay) [inline onnxruntime/core/framework/stream_execution_context.h:40]: Not clear how this is related to enabling the warning

**Review by skottmckay — DISMISSED** [review]

File: cmake/onnxruntime_providers.cmake [hunk 4]
```diff
@@ -1785,6 +1785,12 @@ if (onnxruntime_USE_XNNPACK)
             RUNTIME   DESTINATION ${CMAKE_INSTALL_BINDIR}
             FRAMEWORK DESTINATION ${CMAKE_INSTALL_BINDIR})
   endif()
+
+  # TODO fix shorten-64-to-32 warnings
+  # there are some in builds where sizeof(size_t) != sizeof(int64_t), e.g., in 'ONNX Runtime Web CI Pipeline'
```
Comment (edgchen1) [inline cmake/onnxruntime_providers.cmake:1790]: TODO for a follow up PR, don't want to change too many files in this one

File: onnxruntime/core/framework/stream_execution_context.h (hunk 3 above)
Comment (edgchen1) [reply onnxruntime/core/framework/stream_execution_context.h]: I redid this change. the original change was to make the Get()/Set() types consistent with the atomic's underlying type (int_fast32_t), which was possibly wider than int32_t.

**Review by skottmckay — APPROVED** [review]
:shipit:
